### PR TITLE
Refactor GAIL as subclass of TRPO

### LINKED ISSUE #267: Expose self.sess for GAIL
Unlike all of the other models, the GAIL class does not expose its attribute `self.sess`. Instead, one has to reach into the TRPO attribute and reference `self.trpo.sess`. This leads to displeasing code like:
```python 
if isinstance(model, GAIL):
    model.trpo.sess.close()
else:
    model.sess.close()
```
I propose we expose the attribute with the code below, similar to [this line](https://github.com/hill-a/stable-baselines/blob/master/stable_baselines/gail/model.py#L52)
```python
# In the GAIL constructor
self.sess = self.trpo.sess
```

### PR DESCRIPTION
**Pull request by kantneel**
Accompanies Issue #267 

Closes #267

### CODE REVIEW

**Comment by araffin** [comment]
Btw, I think it is a good idea to get rid of the matplotlib import (that breaks docker installation) in this PR, as suggested by @AdamGleave in https://github.com/HumanCompatibleAI/adversarial-policies/pull/8 ;)

File: stable_baselines/gail/model.py [hunk 1]
```diff
@@ -38,35 +38,41 @@ def __init__(self, policy, env, expert_dataset=None,
                  hidden_size_adversary=100, adversary_entcoeff=1e-3,
                  g_step=3, d_step=1, d_stepsize=3e-4, verbose=0,
                  _init_setup_model=True, **kwargs):
+        self.trpo = TRPO(policy, env, verbose=verbose, _init_setup_model=False, **kwargs)
```
Comment (AdamGleave) [inline stable_baselines/gail/model.py]: Why is this moved to before the super class constructor? (Seems like it'll work OK either way, and in general seems better to have things after the super call.)

**Comment by AdamGleave** [comment]
I agree with my earlier comment to remove the matplotlib import ;) Otherwise this PR looks good to me. @kantneel please ping me when you've finished and I'll review.

File: stable_baselines/gail/model.py (hunk 1 above)
Comment (kantneel) [reply stable_baselines/gail/model.py]: If I have it after the constructor call then `self.trpo` does not yet exist. This means that the magic methods will have to handle the specific case of `self.trpo` not yet existing, which seemed rather unnecessary considering an instance of GAIL must have an instance of TRPO.

Comment (kantneel) [reply stable_baselines/gail/model.py]: Then again, the attributes that are set in the constructor should belong to GAIL and GAIL alone, so I will go ahead and adjust this accordingly.

**Comment by kantneel** [comment]
Okay, I adjusted the matplotlib import and also responded to @AdamGleave's comment.

**Comment by AdamGleave** [comment]
I understand now why you had the super-class constructor later. Unfortunately I think neither are quite right.

In the current implementation, the super-class constructor sets e.g. `self.sess` to `None`. Instance attributes already set override the `__getattr__` method, so if I create a `gail` instance I get `gail.sess` as `None` but `gail.trpo.sess` as non-`None`; same problem as before.

In the previous implementation, the super-class constructor instead reset `self.trpo.sess` to `None` (indirectly). This worked out OK since `TRPO` only set `sess` in `setup_model` that gets called later. But it's somewhat accidental that this works.

I'm now confused why `GAIL` is designed in this way at all; shouldn't it be a subclass of `TRPO` rather than some confusing wrapper? I've pushed a change that does this; @erniejunior and @kantneel can you review?

**Comment by kantneel** [comment]
Outside of the code, it sounds strange to say that GAIL is a specific version of TRPO. However, the way the code is implemented, TRPO is the class which holds all of the GAIL-related attributes anyway, e.g. `d_step`, `g_step`. So I agree, GAIL should just be a subclass of TRPO if we choose to keep the TRPO implementation as it is. Makes the GAIL class and code using it much simpler.

**Comment by araffin** [comment]
>I'm now confused why GAIL is designed in this way at all; shouldn't it be a subclass of TRPO rather than some confusing wrapper?

Very good point. This is again because of legacy reason. If you look at the [original repo](https://github.com/openai/baselines/tree/master/baselines/gail), this is even worse: there is a duplicated version of TRPO. That's what we did at first: unify both TRPO.

So making it a subclass of TRPO is a good change ;)

File: stable_baselines/gail/dataset/dataset.py [hunk 2]
```diff
@@ -155,6 +154,7 @@ def plot(self):
         """
         Show histogram plotting of the episode returns
         """
+        import matplotlib.pyplot as plt
```
Comment (araffin) [inline stable_baselines/gail/dataset/dataset.py]: please add a comment to explain why this is not a top level import ;).

**Comment by araffin** [comment]
@kantneel thanks for you PR, only a small comment to add and it is ready to merge =)
